### ISSUE TITLE
Snowflake / L067: `fix` for L067 mistakes `:` (colon) with double colon

### ISSUE THREAD

**Issue by cbrswt** [sqlfluff/sqlfluff#4453]
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

SQLFluff finds a violation of L067 in `CAST(TRIM(value:Longitude::varchar) AS DOUBLE) as longitude` (where `value` is Snowflake's `VARIANT` type). If you run SQLFluff in *fix* mode with `preferred_type_casting_style` set to `colon`, it outputs a correct result:
```sql
(TRIM(value:Longitude::VARCHAR))::DOUBLE AS longitude
```
However, if set to `cast` it probably mistakes `:` as cast, returning an incorrect SQL:
```sql
CAST(TRIM(CAST(CAST(value AS :Longitude) AS varchar)) AS DOUBLE) as longitude
```
(it should be `value:Longitude` still, not `value AS :Longitude`.

### Expected Behaviour

Fixing `CAST(TRIM(value:Longitude::varchar) AS DOUBLE) as longitude` should return
```sql
CAST(TRIM(CAST(value:Longitude AS varchar) AS DOUBLE) as longitude
```

### Observed Behaviour

```sql
CAST(TRIM(CAST(CAST(value AS :Longitude) AS varchar)) AS DOUBLE) as longitude
```

### How to reproduce

```sql
-- example.sql
{{ config(materialized = "table") }}

SELECT
    CAST(TRIM(value:Longitude::varchar) AS DOUBLE) as longitude
FROM
    (
    	SELECT TO_VARIANT(PARSE_JSON('{"Longitude": "-87.642098091", "Latitude": "41.946275797"}')) AS value
    )
```

```bash
sqlfluff fix models/example.sql
```

### Dialect

Snowflake

### Version

* sqlfluff, version 2.0.0a5
* Python 3.9.15
* dbt Core: 1.4.1
* dbt snowflake: 1.4.1
* sqlfluff-templater-dbt==2.0.0a5

### Configuration

```
[sqlfluff]
templater = dbt
dialect = snowflake
large_file_skip_byte_limit = 0
processes = 0
exclude_rules = L016,L031

[sqlfluff:templater:jinja]
apply_dbt_builtins = True

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower

[sqlfluff:rules:L040]
capitalisation_policy = upper

[sqlfluff:rules:L057]
additional_allowed_characters = "$"

[sqlfluff:rules:L063]
extended_capitalisation_policy = upper

[sqlfluff:rules:L067]
preferred_type_casting_style = cast
```

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

**Comment by justincanney** [comment]
Can anyone confirm if this is fixed in the 2.0.2 release? 
I've just upgraded to 2.0.2 and am still observing the same behavior. 

```
(venv) ~/local-development/ sqlfluff --version
sqlfluff, version 2.0.2
```

sqlfluff fix is returning the incorrect formatting: 
```
SELECT CAST(TRIM(CAST (CAST (value AS:Longitude) AS varchar)) AS double) AS longitude
```
